db_sql.go

```diff
@@ -542,6 +542,64 @@ func (x *sqlUserStoreDB) UnarchiveIdentity(userId UserId) error {
 	return nil
 }
 
+func (x *sqlUserStoreDB) SetUserStats(userId UserId, action string) error {
+	var tx *sql.Tx
+	var err error
+	var update sql.Result
+
+	if tx, err = x.db.Begin(); err != nil {
+		return fmt.Errorf("could not begin transaction: %v", err)
+	}
+	defer tx.Rollback()
+
+	if action == UserStatActionLogin {
+		update, err = tx.Exec(`INSERT INTO authuserstats (user_id, last_login_date)
+							   VALUES ($1, NOW())
+							   ON CONFLICT (user_id) 
+							   DO UPDATE SET last_login_date = NOW()`, userId)
+	}
+	if action == UserStatActionEnable {
+		update, err = tx.Exec(`INSERT INTO authuserstats (user_id, enabled_date, disabled_date)
+							   VALUES ($1, NOW(), NULL)
+							   ON CONFLICT (user_id) 
+							   DO UPDATE SET enabled_date = NOW(), disabled_date = NULL`, userId)
+	}
+	if action == UserStatActionDisable {
+		update, err = tx.Exec(`INSERT INTO authuserstats (user_id, enabled_date, disabled_date)
+							   VALUES ($1, NULL, NOW())
+							   ON CONFLICT (user_id)
+							   DO UPDATE SET enabled_date = NULL, disabled_date = NOW()`, userId)
+	}
+	if err != nil {
+		return fmt.Errorf("could not update auth user stats: %v", err)
+	}
+	if affected, _ := update.RowsAffected(); affected != 1 {
+		return fmt.Errorf("user stats could not be updated: %v", ErrIdentityAuthNotFound)
```

Comment:
This error is not necessarily because the user was not found...it could be because rowsaffected returned more than one row...even though the sql makes that (hopefully) impossible, we should probably make it clearer.